### ISSUE TITLE
PPOTrainer - Potential minor bug In batched_forward_pass()

### ISSUE THREAD

**Issue by samuelarnesen** [huggingface/trl#1066]
The model being trained is set to evaluation mode at the start of the batched_forward_pass() function inside the PPOTrainer class ([reference](https://github.com/huggingface/trl/blob/main/trl/trainer/ppo_trainer.py#L970)). I believe this is appropriate for most invocations of batched_forward_pass(), but not all cases. For example, I think we should be in training mode for the part of the step() function when we are trying to collect gradients for the subsequent backwards pass ([reference](https://github.com/huggingface/trl/blob/main/trl/trainer/ppo_trainer.py#L970)).

For a concrete example of where this could cause issues, I'm trying to use PPOTrainer to train a LlamaModel using AutoModelForCausalLMWithValueHead and gradient checkpointing.  However, the gradient checkpointing function is not being invoked because LlamaModel confirms whether it is in training mode before running gradient checkpointing ([reference](https://github.com/huggingface/transformers/blob/c832bcb812fc962830c11ea64c5ff623240a3d6d/src/transformers/models/llama/modeling_llama.py#L914)). I've been able to circumvent the issue by monkeypatching the batched_forward_pass() function, but ideally I think this can be fixed within the PPOTrainer class.

I'm happy to submit a PR to address this if people think this is an issue worth addressing. 

Thank you!

**Comment by vwxyzjn** [comment]
Hi @samuelarnesen, thanks for raising this issue. The reason for running `model.eval()` is to set `dropout` probabilities to be zero, so we get reliable / reproducible logprobs. Here is an example:

```python
import torch
import transformers
import torch.nn.functional as F
tokenizer = transformers.AutoTokenizer.from_pretrained("gpt2", padding_side="right")
tokenizer.add_special_tokens({"pad_token": "[PAD]"})
pad_id = tokenizer.pad_token_id
query = torch.tensor([
    [23073],
])
response = torch.tensor([
    [11, 339, 561],
])
context_length = query.shape[1]
query_response = torch.cat((query, response), 1)
model = transformers.AutoModelForCausalLM.from_pretrained("gpt2")


def get_logprobs(query_response):
    output = model(input_ids=query_response)
    logits = output.logits
    all_logprobs = F.log_softmax(logits, dim=-1)
    logprobs = torch.gather(all_logprobs, 2, response.unsqueeze(-1)).squeeze(-1)
    return logprobs

model.eval()
initial_logprobs = get_logprobs(query_response)
print(f"{initial_logprobs=}")

print("model.eval()")
model.eval()
training_logprobs = get_logprobs(query_response)
print(f"{training_logprobs=}")
kl_divergence = training_logprobs - initial_logprobs
print(f"{kl_divergence=}")

print("model.train()")
model.train()
training_logprobs = get_logprobs(query_response)
print(f"{training_logprobs=}")
kl_divergence = training_logprobs - initial_logprobs
print(f"{kl_divergence=}")
```

```
(trl-py3.9) ➜  trl git:(fix-reward-gather) ✗ python -i causalllm4.py
initial_logprobs=tensor([[-3.3213, -4.9980, -3.8690]], grad_fn=<SqueezeBackward1>)
model.eval()
training_logprobs=tensor([[-3.3213, -4.9980, -3.8690]], grad_fn=<SqueezeBackward1>)
kl_divergence=tensor([[0., 0., 0.]], grad_fn=<SubBackward0>)
model.train()
training_logprobs=tensor([[-3.5822, -5.0551, -3.6366]], grad_fn=<SqueezeBackward1>)
kl_divergence=tensor([[-0.2609, -0.0571,  0.2324]], grad_fn=<SubBackward0>)
```

As you can see, the KL divergence is much larger than expected since the model is the same. For this reason, it will cause other issues (e.g., KL penalty being larger than what it should be, causing PPO's ratio to go significantly beyond 1 and therefore training instability).

A possible way to still use `model.train()` is to manually set the dropout probs to zero, like done in https://github.com/microsoft/DeepSpeedExamples/blob/737c6740bec38b77a24a59135b6481a53d566b38/applications/DeepSpeed-Chat/training/utils/model/model_utils.py#L20C1-L26C52

```python
def configure_dropout(model_config, dropout_layer_keys, dropout):
    if dropout is not None:
        for key in dropout_layer_keys:
            if hasattr(model_config, key):
                print(f"Setting model_config.{key} to {dropout}")
                setattr(model_config, key, dropout)

model_config = transformers.AutoConfig.from_pretrained("gpt2")
configure_dropout(model_config, ["attn_pdrop", "embd_pdrop", "resid_pdrop", "summary_first_dropout"], 0.0)  # disable dropout
model = transformers.AutoModelForCausalLM.from_pretrained("gpt2", config=model_config)
print("model.train() w/ dropout disabled")
model.train()
training_logprobs = get_logprobs(query_response)
print(f"{training_logprobs=}")
kl_divergence = training_logprobs - initial_logprobs
print(f"{kl_divergence=}")
```
```
Setting model_config.attn_pdrop to 0.0
Setting model_config.embd_pdrop to 0.0
Setting model_config.resid_pdrop to 0.0
Setting model_config.summary_first_dropout to 0.0
model.train() w/ dropout disabled
training_logprobs=tensor([[-3.3213, -4.9980, -3.8690]], grad_fn=<SqueezeBackward1>)
kl_divergence=tensor([[0., 0., 0.]], grad_fn=<SubBackward0>)
```

While this approach is probably more desirable because it enables us to leverage the train mode of `LayerNorm`, we would need to maintain a list of dropout keys for all the models we support. In practice I haven't seen much of a different in smaller models between `model.eval()` and `configure_dropout`, but maybe larger models this could become more important.

CC @lvwerra @lewtun @younesbelkada

**Comment by samuelarnesen** [comment]
Thanks so much for the response! The downstream effect on features like gradient checkpointing is unfortunate since it's causing out-of-memory errors but it sounds like there's no solution here that exists without meaningful tradeoffs. I'll mark this as closed. Thanks again!
